mprotect_builder.h

```diff
@@ -0,0 +1,62 @@
+#pragma once
+
+#include <stdint.h>
+#include <vector>
+
+#include "utils.h"
+
+class MprotectBuilder {
+public:
+    void Add(uintptr_t offset, uintptr_t size) {
+        offsets.emplace_back(offset);
+        sizes.emplace_back(size);
+    }
+    uintptr_t Size() const { return sizeof(memprotect_body_code) * offsets.size() + sizeof(memprotect_end_code); }
+    void Emit(FILE* fp, uintptr_t mprotect_code_offset) {
+        uint8_t* mprotect_code = (uint8_t*)malloc(Size());
+        uint8_t* mprotect_code_head = mprotect_code;
+        for (int i = 0; i < offsets.size(); i++) {
+            LOG(INFO) << "MprotectBuilder::Emit: " << SOLD_LOG_BITS(offsets[i]) << SOLD_LOG_BITS(sizes[i]);
+
+            memcpy(mprotect_code_head, memprotect_body_code, sizeof(memprotect_body_code));
+            int64_t* offset_p = (int64_t*)(mprotect_code_head + memprotect_body_addr_offset);
+            // 17 is the length of `mov $0xdeadbeefdeadbeef, %rdi; lea (%rip), %rsi'
+            int64_t offset_v = (offsets[i] & (~(0x1000 - 1))) -
+                               (static_cast<int64_t>(mprotect_code_offset) + static_cast<int64_t>(mprotect_code_head - mprotect_code) + 17);
+            LOG(INFO) << "MprotectBuilder::Emit" << SOLD_LOG_BITS(offset_v) << SOLD_LOG_KEY(offset_v);
+            *offset_p = offset_v;
+            uint32_t* size_p = (uint32_t*)(mprotect_code_head + memprotect_body_size_offset);
+            *size_p = ((offsets[i] + sizes[i]) & (~(0x1000 - 1))) - (offsets[i] & (~(0x1000 - 1)));
+            mprotect_code_head += sizeof(memprotect_body_code);
+        }
+        memcpy(mprotect_code_head, memprotect_end_code, sizeof(memprotect_end_code));
+        mprotect_code_head += sizeof(memprotect_end_code);
+
+        CHECK(fwrite(mprotect_code, sizeof(uint8_t), Size(), fp) == Size());
+        free(mprotect_code);
+    }
+
+    // call SYS_mprotect syscall
+    //
+    // mov $0xdeadbeefdeadbeef, %rdi
+    // lea (%rip), %rsi
+    // add %rsi, %rdi
+    // mov $0xaabbcc, %rsi (size)
+    // mov $0x1, %rdx (0x1 = PROT_READ)
+    // mov $10, %eax (10 = SYS_mprotect)
+    // syscall
```

Comment:
Better to execute `ud2` or something to let the program crash when `mprotect` fails? I think the code should be like

```
    syscall
    test %eax, %eax
    jz ok
    ud2
ok:
```

Reply:
It looks good!! I adopted it in https://github.com/shinh/sold/pull/59/commits/ab0bfa96f154dd3edc78363df7f6d3c868536d05.